# Sessió de problemes 5: Recursivitat, teorema Màster i fitxers

## Problema 1: Força bruta: producte de conjunts

Donats dos conjunts, $A$ i $B$, el producte d'aquests dos conjunts és el conjunt $A \times B$ format per totes les parelles ordenades $(a, b)$ en què $a \in A$ i $b \in B$.

Si tenim un nombre conegut de conjunts és fàcil fer-ho amb comprehensions:

In [8]:
{a+b+c for a in {'A','B','C'} for b in {'D','E','F'} for c in {'G','H'}}
#En aquest set, no es podran repetir les igualtats.

{'ADG',
 'ADH',
 'AEG',
 'AEH',
 'AFG',
 'AFH',
 'BDG',
 'BDH',
 'BEG',
 'BEH',
 'BFG',
 'BFH',
 'CDG',
 'CDH',
 'CEG',
 'CEH',
 'CFG',
 'CFH'}

Però si volem fer una funció que calculi el producte d'un nombre variable de conjunts, ja no es tant fàcil fer-ho amb comprensions i ho és més fer-ho recursivament. Com ho farieu?

In [ ]:
'''
El problema és que no es sap la quantitat que hi ha.
Quan n'hi han dos, no hi ha problema, pero quan en tenim +2, haurem de fer recursivitat.

'''
a = {'A','B','C'} 
b = {'D','E','F'} 
c = {'G','H'}
l1 = [a,b,c]

def set_prod(l):
    pass

assert set_prod(l1) == {'ADG', 'ADH', 'AEG', 'AEH', 'AFG', 'AFH', 'BDG', 'BDH', 'BEG', 'BEH', 'BFG', 'BFH', 'CDG', 'CDH', 'CEG', 'CEH', 'CFG', 'CFH'}

assert set_prod(l2) == {'AD', 'AE', 'AF', 'BD', 'BE', 'BF', 'CD', 'CE', 'CF'}

## Problema 2: Combinacions de tecles

<img src="keypad.png" width="250px" alt="2 ABC, 3 DEF, 4 GHI, 5 JKL, 6 MNO, 7 PQRS, 8 TUV, 9 WXYZ" style="float:left; margin:0 10px 0 0" /> 


Donada una seqüència de nombres entre 2 i 9, mostrar totes les possibles paraules que es poden formar amb les lletres associades a aquests nombres en un teclat de mòbil antic, com el que està dibuixat aquí.


<div style="font-style:italic">
Per exemple:
<ul>
<li>Tecles=[tecla2,tecla3,tecla4]</li>
<li>Solució:
   <ul>
   <li>ADG, BDG, AEG, BEG, CEG, AFG, BFG, CFG, ADH, BDH, CDH, AEH, BEH, CEH, AFH, BFH, CFH, ADI, BDI, CDI, AEI, BEI, CEI, AFI, BFI, CFI</li>
   </ul>
</li>
</ul>
</div>

In [13]:
tecles=[]
fitxer=open("tecles.txt","r")

for i in fitxer:
    tecles.append(i.replace("\n","").split(","))
fitxer.close()
        
print(tecles)
    
    



[['A', 'B', 'C'], ['D', 'E', 'F'], ['G', 'H', 'I'], ['J', 'K', 'L'], ['M', 'N', 'O'], ['P', 'Q', 'R', 'S'], ['T', 'U', 'V'], ['W', 'X', 'Y', 'Z']]


Escriviu una versió de la solució en la que llegiu les lletres associades a cada tecla des del fitxer "tecles.txt" que us facilitem.

In [26]:
def combinacions(teclat):
    if len(teclat)==0:
        return ""
    elif len(teclat)==1:
        return teclat[0]
    elif len(teclat)==2:
        solucio={a+b for a in teclat[0] for b in teclat[1]}
        return solucio
    else:
        return combinacions([teclat[0],combinacions(teclat[1:])])

In [27]:
teclat=[tecles[0],tecles[3],tecles[4]] 
combinacions(teclat)

{'AJM',
 'AJN',
 'AJO',
 'AKM',
 'AKN',
 'AKO',
 'ALM',
 'ALN',
 'ALO',
 'BJM',
 'BJN',
 'BJO',
 'BKM',
 'BKN',
 'BKO',
 'BLM',
 'BLN',
 'BLO',
 'CJM',
 'CJN',
 'CJO',
 'CKM',
 'CKN',
 'CKO',
 'CLM',
 'CLN',
 'CLO'}

#### Pregunta 2.1: Quantes combinacions es poden crear amb 3 tecles? Quines combinacions es poden crear amb [2,3]?

Depèn de quantes n'agafem. 

#### Pregunta 2.2: Si $N$ és el nombre de tecles d'entrada, quina és la complexitat d'aquest algorisme?
**Atenció**: abans de respondre comproveu si l'algorisme compleix els requisits per usar el teorema del màster, en cas contrari feu el càlcul de complexitat per combinatòria


No es pot respondre, ja que no compleix el teorema master. 
Seria una complexitat de 4^n, ja que mirem les maximes combinacions possibles i més grans.

## Problema 3: Càlcul de complexitat

#### Pregunta 3.1. Calcula la complexitat de la següent funció recursiva, considerant la mida $n$ dels nombres en bits.

In [14]:
def multiplica(x, y):
    """
    Aquesta funció retorna la multiplicació de dos nombres expressats en binari
    :param: x,y són nombres binaris introduits com a cadenes de text
    :return: el resultat de la multiplicació
    """
    n = max(len(x), len(y))
    if n == 1:
        return int(x)*int(y)
    xleft, xright = x[:n//2], x[n//2:]
    yleft, yright = y[:n//2], y[n//2:]
    p1 = multiplica(xleft, yleft)
    p2 = multiplica(xright, yright)
    p3 = multiplica(str(int(xleft)+int(xright)), str(int(yleft)+int(yright)))
    return p1*2**n+(p3-p1-p2)*2**(n/2)+p2

multiplica("1111","1100") 

180.0

Solució

a) Apuntem les complexitats de les operacions simples: 

| Operació | Complexitat |
| -------- | ----------- |
|   max()  |  O(1) |
| len() | O(1) | 
| slice | O(k)=O(n/2) | 
| int(str) | O(n)  |
| str(int) | O(n) |
                  

    
b) Apuntem la complexitat de cada bloc de codi

| Bloc | Complexitat |
| -------- | ----------- |
|  n = max(len(x), len(y))        |    O(2)=>O(1)     |
|  | ----------- |

     
c) Determinem els paràmetres del teorema de màster:  a =  , b =  , d =  . 
       
d) Identifiquem el cas pertinent (cas  ) i apliquem el teorema de màster: O( ).
    
La funció multiplica té un cost      respecte al nombre de bits del nombre més llarg.
    

#### Pregunta 3.2 Calcula la complexitat segons el teorema de màster:

<ol style="list-style-type:lower-alpha">
<li>$T(n)=2T(n/2)+n^{4}$</li>
<li>$T(n)=T(7n/10)+n$</li>
<li>$T(n)=16T(n/4)+n^{2}$</li>
<li>$T(n)=2T(n/4)+sqrt(n)$</li>
</ol>

## Problema 4: Quantitat d'1s a una llista binària

Donada una llista de nombres binaris ordenada, trobar de manera eficient el nombre d'1s que conté.

Per exemple:
+ La ``llista=[0,0,1,1,1,1,1]`` conté 5 1s.
+ La ``llista=[0,0,0,0,1,1]`` conté 2 1s.

In [38]:
# Inserir aquí el codi de la solució

def Nombre1(llista):
    """
    Aquesta funció, donada una llista de nombres binaris ordenada,
    troba de manera eficient el nombre d'1s que conté
    :param llista, la llista de nombres binaris ordenada
    :return el nombre d'1s que conté
    """
    if len(llista)==0:
        return 0
    elif llista[0]==1:
        return len(llista)
    elif llista[-1]==0:
        return 0
    else:
        mid= (len(llista))//2
        if mid==0:
            return Nombre1(llista[mid:])
        else:
            return Nombre1(llista[:mid])+Nombre1(llista[mid:])
    
    

In [39]:
Nombre1([0,0,1,1,1,1,1])

5

In [3]:
def recbinsearch(k, nums, low, high):
# low, high defineixen els limits de la llista
# on buscar, així no cal crear noves llistes
    if low>high:
        return -1
    mid = (low + high) // 2
    items = nums[mid]
    if items == k:
        return mid
    elif k < items:
        return recbinsearch(k,nums,low,mid-1)
    else:
        return recbinsearch(k,nums,mid+1,high)

In [4]:
recbinsearch(3,[1,2,3,4,5,6,7,8,9], 0, 8)

2

#### Pregunta 4.1 Si plantegem el problema com un algorisme recursiu, quins serien els casos base? Observa els següents casos:

- Llista:``[]``
- Llista:``[0,0]``
- Llista:``[1,1,1,1,1,1]``
- Llista:``[0,1,1]``

*Indica aquí la teva solució*

#### Pregunta 4.2 Calcula la complexitat de l'algorisme pel teorema màster

*Indica aquí la teva solució*

## Problema 5: Elements Pic

Donada una llista d’enters, trobar un dels elements *pic*. Un enter serà un element *pic* si els seus veïns immediats són menors que ell. 

<div style="font-style:italic">
Per exemple:
<ul>
<li>Llista=[3,1,5,2,7,8]</li>
<li>Solució:
   <ul>
   <li>Un dels elements pic és el 3</li>
   <li>En aquesta llista en realitat hi ha diversos elements pic: el 3, el 5 i el 8.</li>
   </ul>
</li>
<li>Llista=[9,5,2]</li>
<li>Solució:
   <ul>
   <li>Un dels elements pic és el 9</li>
   <li>En una llista ordenada en ordre descendent, l'element pic sempre serà el primer enter.</li>
   </ul>
</li>
<li>Llista=[1,2,7,8]</li>
<li>Solució:
   <ul>
   <li>Un dels elements pic és el 8</li>
   <li>En una llista ordenada en ordre ascendent, l'element pic sempre serà el darrer enter.</li>
   </ul>
</li>
</ul>
</div>

In [1]:
def elementPic(llista):
    """
Donada una llista d’enters, trobar un dels elements pic.
Un enter serà un element pic si els seus veïns immediats són menors que ell.
:param llista, la llista d'enters
:return els elements pics o -1 si no en troba
"""
    i=0
    if llista[i]<llista[i+1]:
        i+=1
    else:
        return ll

            
        
    
    

assert elementPic([1,2,7,8]) == 8


AssertionError: 

In [2]:
#elementPic([1,2,7,8])
elementPic([1,2,7,5,6,7])


#### Pregunta 5.1: Si plantegem el problema com un algorisme recursiu. Quins són els casos base? Observeu les següents llistes:

- ```llista1 = [8,9,10,2,5,6]```
- ```llista2 = [8,9,10,12,15]```
- ```llista3 = [10,8,6,5,3,2]```

Els que estan envoltats de números menors....

#### Pregunta 5.2: Calcula la complexitat de l'algorisme amb el teorema màster.

*Indica aquí la teva solució*

## Problema 6: Nombres estrictament creixents

<p style="font-family:Arial;font-size:1em">
Donat un enter N entre 1 i 9, trobar tots els nombres amb N dígits, dels quals el valor és estrictament creixent.
</p>

*Per exemple*:
+ N=8
+ Solució: 12345678, 12345679, 12345689, 12346789, 12356789, 12456789, 13456789, 23456789

In [9]:
def estrCreix(n):
    """
Aquesta funció troba tots els nombres amb N dígits, amb valor
estrictament creixent
:param n, un enter entre 1 i 9
:return no retorna res, mostra el resultat per pantalla
"""
#S'haurà de anar canviat un nombre, però sempre tindrem en la primera posició l'1 i el 2
    estrCreixRec(str(''),n,0)
    

    
    


def estrCreixRec(cadena, n, previ):
    """
Aquesta funció troba tots els nombres amb N dígits, amb valor
 estrictament creixent
:param cadena, una cadena amb la preparació d'un nombre creixent amb x dígits
:param n, un enter entre 1 i 9, que en aquest cas indica els dígit que falta
          completar
:param previ, el darrer dígit de la cadena
:return no retorna res, mostra el resultat per pantalla
"""
    if n==0:
        print(cadena)
    
    for i in range(previ+1,(10-n+1)%11):
        estrCreixRec(cadena+str(i),n-1,i)
            
            
    



In [13]:
estrCreix(9)

123456789


#### Pregunta 6.1: Quina solució hauria de donar per als següents valors:

- N=9
- N=2

N=9 123456789
Per a n=2
12
13
14
15
16
17
18
19
23
24
25
26
27
28
29
34
35
36
37
38
39
45
46
47
48
49
56
57
58
59
67
68
69
78
79
89

#### Pregunta 6.2: Imaginem que volem construir un número de 7 dígits (N=7) i que ja tenim escrit el nombre fins al dígit 6è (123456). Quantes crides recursives ha de fer el programa encara?

Ha de fer la crida que correspon al 7,8 i 9: 3 crides

#### Pregunta 6.3: Totes les crides recursives generaran cadenes correctes? se t'acudeix alguna millora?

Si, crec que el programa no dona errors

#### Pregunta 6.4: La complexitat d'aquest algorisme es pot calcular pel teorema del màster? per què?